truffle/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/SourceSectionFilter.java

```diff
@@ -0,0 +1,540 @@
+ * used in an Truffle interpreter.
+ * <p>
+ * Start building event filters by calling {@link SourceSectionFilter#newBuilder()} and complete
+ * them by calling {@link Builder#build()}.
+ *
+ * @see SourceSectionFilter#newBuilder()
+ * @see Instrumenter#attachFactory(SourceSectionFilter, EventNodeFactory)
+ * @see Instrumenter#attachListener(SourceSectionFilter, EventListener)
+ */
+public final class SourceSectionFilter {
+
+    private final EventFilterExpression[] nodeExpressions;
+    private final EventFilterExpression[] rootNodeExpressions;
+
+    private SourceSectionFilter(EventFilterExpression[] rootNodeExpressions, EventFilterExpression[] nodeExpressions) {
+        this.rootNodeExpressions = rootNodeExpressions;
+        this.nodeExpressions = nodeExpressions;
+    }
+
+    /**
+     * Creates a new {@link SourceSectionFilter} expression using a {@link Builder builder} pattern.
+     * Individual builder statements are interpreted as conjunctions (AND) while multiple parameters
+     * for individual filter expressions are treated as disjunctions (OR). To create the final
+     * filter finalize the expression using {@link Builder#build()}.
+     *
+     * @see Builder#sourceIs(Source...)
+     * @see Builder#mimeTypeIs(String...)
+     * @see Builder#tagIs(String...)
+     * @see Builder#tagIsNot(String...)
+     * @see Builder#sourceSectionEquals(SourceSection...)
+     * @see Builder#indexIn(int, int)
+     * @see Builder#lineIn(int, int)
+     * @see Builder#lineIs(int)
+     * @see Builder#build()
+     *
+     * @return a new builder to create
+     */
+    public static Builder newBuilder() {
+        return new Builder();
+    }
+
+    public static final class Builder {
+
+        private List<EventFilterExpression> nodeExpressions = new ArrayList<>();
+        private List<EventFilterExpression> rootNodeExpressions = new ArrayList<>();
+
+        /**
+         * Add a filter for all source sections that reference one of the given sources.
+         */
+        public Builder sourceIs(Source... source) {
+            verifyNotNull(source);
+            rootNodeExpressions.add(new EventFilterExpression.SourceIs(source));
+            return this;
+        }
+
+        /**
+         * Add a filter for all source sections that declare one of the given mime-types.
+         */
+        public Builder mimeTypeIs(String... mimeTypes) {
+            verifyNotNull(mimeTypes);
+            rootNodeExpressions.add(new EventFilterExpression.MimeTypeIs(mimeTypes));
+            return this;
+        }
+
+        /**
+         * Add a filter for all source sections that are tagged with one of the given String tags.
+         */
+        public Builder tagIs(String... tags) {
+            verifyNotNull(tags);
+            nodeExpressions.add(new EventFilterExpression.TagIs(tags));
+            return this;
+        }
+
+        /**
+         * Add a filter for all sources sections that declare not one of the given String tags.
```

Comment:
What happens if: a builder's  expressions include `….tagIs(“doesNotCompute”)…..tagIsNot(“doesNotCompute”)…`?  The specification tells me that the filter will never match anything; can you optimize this?  I hope it doesn’t create a paradox that causes the computer to explode :-)